stumpy/scrump.py

```diff
@@ -103,67 +109,111 @@ def _compute_PI(
     for i in indices[start:stop]:
         Q = T_A[i : i + m]
         QT[:] = core._sliding_dot_product(Q, T_B)
-        # Update P[i] relative to all T[j : j + m]
         squared_distance_profile[:] = core._mass(Q, T_B, QT, μ_Q[i], σ_Q[i], M_T, Σ_T)
         squared_distance_profile[:] = np.square(squared_distance_profile)
         if excl_zone is not None:
-            zone_start = max(0, i - excl_zone)
-            zone_stop = min(l, i + excl_zone)
-            squared_distance_profile[zone_start : zone_stop + 1] = np.inf
-
-            # only for self-join
-            mask = squared_distance_profile < P_squared[thread_idx]
-            P_squared[thread_idx][mask] = squared_distance_profile[mask]
-            I[thread_idx][mask] = i
-
-        I[thread_idx, i] = np.argmin(squared_distance_profile)
-        P_squared[thread_idx, i] = squared_distance_profile[I[thread_idx, i]]
-        if P_squared[thread_idx, i] == np.inf:  # pragma: no cover
-            I[thread_idx, i] = -1
-        else:
-            j = I[thread_idx, i]
-            # Given the squared distance, work backwards and compute QT
-            QT_j = (m - P_squared[thread_idx, i] / 2.0) * (Σ_T[j] * σ_Q[i]) + (
-                m * M_T[j] * μ_Q[i]
+            core._apply_exclusion_zone(squared_distance_profile, i, excl_zone, np.inf)
+
+        # find EXACT (not approx.) value of `P_squared[thread_idx, i, 0]`
+        nn = np.argmin(squared_distance_profile)
+        core._shift_insert_at_index(
+            P_squared[thread_idx, i], 0, squared_distance_profile[nn]
+        )
+        core._shift_insert_at_index(I[thread_idx, i], 0, nn)
+
+        if P_squared[thread_idx, i, 0] == np.inf:  # pragma: no cover
+            I[thread_idx, i, 0] = -1
+            continue
```

Comment:
> if the distance is `np.inf` then none of the following "derived" distances will matter either since they will also be `np.inf`. Is that right?

Right! That was the idea that if the smallest distance in the `distance_profile` (of subsequence with start index `i`, `S_i`) is `np.inf`, we can just forget about the rest and go the next iteration (We cannot find the NN index of the `S_i`. Also, we cannot update TopK of other indices using `distance_profile` values).

Reply:
> Have you updated `naive.scrump` accordingly to match this behavior?

By "match this behavior", did you mean `naive.scrump` should follow the same underlying structure?

Reply:
Well, the order in which things are processed may affect the outputs. So I wanted to make sure that they produce the same results and is consistent

Reply:
Ah! Got it! I will check that out!